cheroot/connections.py

```diff
@@ -0,0 +1,254 @@
+            """Stub inheritance prevention.
+
+            Dummy function, since neither fcntl nor ctypes are available.
+            """
+            pass
+    else:
+        def prevent_socket_inheritance(sock):
+            """Mark the given socket fd as non-inheritable (Windows)."""
+            if not _SetHandleInformation(sock.fileno(), 1, 0):
+                raise WinError()
+else:
+    def prevent_socket_inheritance(sock):
+        """Mark the given socket fd as non-inheritable (POSIX)."""
+        fd = sock.fileno()
+        old_flags = fcntl.fcntl(fd, fcntl.F_GETFD)
+        fcntl.fcntl(fd, fcntl.F_SETFD, old_flags | fcntl.FD_CLOEXEC)
+
+
+class ConnectionManager:
+    """Class which manages HTTPConnection objects.
+
+    This is for connections which are being kept-alive for follow-up requests.
+    """
+
+    def __init__(self, server):
+        """Initialize ConnectionManager object.
+
+        Args:
+            server (cheroot.server.HTTPServer): web server object
+                that uses this ConnectionManager instance.
+        """
+        self.server = server
+        self.connections = OrderedDict()
+
+    def put(self, conn):
+        """Put idle connection into the ConnectionManager to be managed.
+
+        Args:
+            conn (cheroot.server.HTTPConnection): HTTP connection
+                to be managed.
+        """
+        self.connections[conn] = time.time(), conn.rfile.has_data()
+
+    def expire(self):
+        """Expire least recently used connections.
+
+        This happens if there are either too many open connections, or if the
+        connections have been timed out.
+
+        This should be called periodically.
+        """
+        if not self.connections:
+            return
+
+        # Check for too many open connections.
+        conns = list(self.connections.items())
+        ka_limit = self.server.keep_alive_conn_limit
+        if ka_limit is not None:
+            [self._close(conn[0]) for conn in conns[:-ka_limit]]
+            conns = conns[-ka_limit:]
+
+        # Check for old connections.
+        now = time.time()
+        for (conn, (ctime, _)) in conns:
+            # Oldest connection out of the currently available ones.
+            if (ctime + self.server.timeout) < now:
+                self._close(conn)
+            else:
+                break
+
+    def get_conn(self, server_socket):
+        """Return a HTTPConnection object which is ready to be handled.
+
+        A connection returned by this method should be ready for a worker
+        to handle it. If there are no connections ready, None will be
+        returned.
+
+        Any connection returned by this method will need to be `put`
+        back if it should be examined again for another request.
+
+        Args:
+            server_socket: ServerSocket instance to listen to for new
+            connections.
+        Returns:
+            cheroot.server.HTTPConnection instance, or None.
+
```

Comment:
```suggestion
```

Reply:
And I did initially have no whitespace there, but then [this CI job](https://travis-ci.org/cherrypy/cheroot/jobs/529425494) failed because of it.

Is there some tool that we can use to deal with auto-format these things prior to commit?

Reply:
Yes. I suggest using [pipx](https://pypi.org/project/pipx) to install [pre-commit](https://pypi.org/project/pre-commit), then `pre-commit install`. This will run pre-commit before committing code. Alternately, for a lighter touch, just run `tox -e pre-commit`, which will run pre-commit manually. Sometimes, pre-commit will automatically edit the files. Other times, it can only raise an error requiring manual intervention.